Pcap++/src/PcapLiveDevice.cpp

```diff
@@ -1056,33 +1063,104 @@ void PcapLiveDevice::setDefaultGateway()
 			PCPP_LOG_ERROR("Error retrieving default gateway address: " << e.what());
 		}
 	}
-#elif defined(__APPLE__) || defined(__FreeBSD__)
-	std::string command = "netstat -nr | grep default | grep " + m_Name;
-	std::string ifaceInfo = executeShellCommand(command);
-	if (ifaceInfo == "")
-	{
-		PCPP_LOG_DEBUG("Error retrieving default gateway address: couldn't get netstat output");
+#elif defined(__APPLE__)
+
+	//route message struct for communication in APPLE device
+	struct BSDRoutingMessage{
+		struct	rt_msghdr header;
+		char	messageSpace[512];
+	};
+
+	struct BSDRoutingMessage routingMessage;
+	char* spacePtr = routingMessage.messageSpace;
+	// It creates a raw socket that can be used for routing-related operations
+	int sockfd = socket(PF_ROUTE, SOCK_RAW, 0);
+	if (sockfd < 0) {
+		PCPP_LOG_ERROR("Error retrieving default gateway address: couldn't get open routing socket");
+		return ;
+	}
+	memset(reinterpret_cast<char*>(&routingMessage), 0, sizeof(routingMessage));
+	routingMessage.header.rtm_msglen = sizeof(struct rt_msghdr);
+	routingMessage.header.rtm_version = RTM_VERSION;
+	routingMessage.header.rtm_type = RTM_GET;
+	routingMessage.header.rtm_addrs = RTA_DST | RTA_NETMASK ;
+	routingMessage.header.rtm_flags = RTF_UP | RTF_GATEWAY | RTF_STATIC;
+
+	routingMessage.header.rtm_msglen += 2*sizeof(sockaddr_in) ;
+
+	if (write(sockfd, reinterpret_cast<char*>(&routingMessage), routingMessage.header.rtm_msglen) < 0) {
+		PCPP_LOG_ERROR("Error retrieving default gateway address: couldn't write into the routing socket");
 		return;
 	}
 
-	// remove the word "default"
-	ifaceInfo.erase(0, 7);
-
-	// remove spaces
-	while (ifaceInfo.at(0) == ' ')
-		ifaceInfo.erase(0,1);
-
-	// erase string after gateway IP address
-	ifaceInfo.resize(ifaceInfo.find(' ', 0));
+	// Read the response from the route socket
+	if (read(sockfd, reinterpret_cast<char*>(&routingMessage), sizeof(routingMessage)) < 0) {
+		PCPP_LOG_ERROR("Error retrieving default gateway address: couldn't read from the routing socket");
+		return;
+	}
 
+	struct in_addr  *gateAddr = nullptr;
+	struct sockaddr *sa = nullptr;
+	spacePtr = (reinterpret_cast<char*>(&routingMessage.header+1));
+    auto rtmAddrs = routingMessage.header.rtm_addrs;
+    int index = 0;
+    auto roundUpClosetMultiple = [](int multiple, int num){
+        return ((num+multiple-1)/multiple)*multiple;
+    };
+    while (rtmAddrs)
+    {
+        if (rtmAddrs & 1) {
+            sa =  reinterpret_cast<sockaddr *>(spacePtr);
+            if(index == RTA_GATEWAY) {
+                gateAddr = internal::sockaddr2in_addr(sa);
+                break;
+            }
+            spacePtr += sa->sa_len > 0 ? roundUpClosetMultiple(sizeof(uint32_t), sa->sa_len) : sizeof(uint32_t);
+        }
+        index++;
+        rtmAddrs >>= 1;
+    }
```

Comment:
Well, according to the CI, this has some issue with `sockaddr2in_addr` being passed a non-IPv4 sockaddr structure.

Reply:
It should be caused by wrongly set index = 0 instead 1.Fixed.